tmt/steps/provision/mrack.py

```diff
@@ -1494,6 +1690,140 @@ def reboot(
         )
 
 
+class BootcImageBuilder:
+    """Separate class to handle bootc image building logic"""
+
+    def __init__(self, data: BeakerGuestData, workdir: Path, logger: tmt.log.Logger):
+        self.data = data
+        self.workdir = workdir
+        self._logger = logger
+
+    def _handle_image_tag(self) -> None:
+        if not self.data.image_tag:
+            image_tag = get_quay_repo_tag(self._logger)
+            assert image_tag
+            self.data.image_tag = image_tag.strip()
+
+    def _get_base_image(self) -> str:
+        """
+        Get base image
+        """
+        if self.data.container_file:
+            return self._build_base_image(self.data.container_file)
+        return self.data.base_image_url
+
+    def build_and_push_image(self) -> str:
+        """Main orchestration method for image building"""
+        if self.data.customize_image:
+            self._handle_image_tag()
+            return self._build_custom_image()
+        return self._use_existing_image()
+
+    def _build_custom_image(self) -> str:
+        """Build custom bootc image"""
+        base_image = self._get_base_image()
+        built_image = self._build_derived_image(base_image)
+        self._push_image(built_image)
+        return f"{self.data.base_repo}/{built_image}"
+
+    def _push_image(self, containerimage: str) -> None:
+        tmt.utils.ShellScript(
+            f"podman login -u {self.data.bootc_user} -p {self.data.bootc_password} "
+            f"{self.data.base_repo}"
+        ).to_shell_command().run(cwd=self.workdir, logger=self._logger)
+        tmt.utils.ShellScript(
+            "podman push --tls-verify=false --quiet "
+            f"{containerimage} {self.data.base_repo}/{containerimage}"
+        ).to_shell_command().run(cwd=self.workdir, logger=self._logger)
+
+    def _use_existing_image(self) -> str:
+        """Use pre-existing bootc image"""
+        if not self.data.image_url:
+            raise tmt.utils.ProvisionError("image_url required when not customizing image")
+        return self.data.image_url
+
+    def _build_base_image(self, container_file: str) -> str:
+        """
+        Build the base container image
+        """
+
+        self._logger.debug("Build container image.")
+        container_workdir = tmt.utils.expand_path(self.data.container_file_workdir)
+        image_name = f'localhost/{self.data.test_image_name}:{self.data.distro_name}'
+        tmt.utils.Command(
+            "podman",
+            "build",
+            container_workdir,
+            "-f",
+            container_file,
+            "-t",
+            image_name,
+        ).run(
+            cwd=self.workdir,
+            stream_output=True,
+            logger=self._logger,
+        )
+        return image_name
+
+    def _build_derived_image(self, containerimage: str) -> str:
+        """
+        Build the container image
+        """
+        platform_dict = {
```

Comment:
All these - `platform_dict` and two templates - should be global, it will be much easier to find them when the time of change comes.

Also, `platform_dict` 1. includes type in its name, which is a code smell, and 2. doesn't say much about its purpose. A better name would be nice.